scipy/integrate/_quadrature.py

```diff
@@ -47,6 +49,20 @@ class AccuracyWarning(Warning):
     pass
 
 
+if TYPE_CHECKING:
+    # workaround for mypy function attributes see:
+    # https://github.com/python/mypy/issues/2087#issuecomment-462726600
+    from typing import Protocol, Dict, Tuple, Any
```

Comment:
So I'd import everything except `Protocol` from typing, and then import `Protocol` from `typing_extensions`. Then instead of inlining `cache_decorator` here I'd probably do

```
if TYPE_CHECKING:
    class CacheAttributes(Protocol):
        cache: Dict[int, Tuple[Any, Any]]
else:
    CacheAttributes = Any


def cache_decorator(func: Callable) -> CacheAttributes:
    ...
```

Reply:
Done